### DISCUSSION TITLE
Custom widget, set selected device programatically

### DISCUSSION THREAD

**Discussion by SindreLindahl** [thingsboard/thingsboard#11889] [Q&A]
How to set a selected device in javascript for custom widget? Have been using two-way command:
`ctx.controlApi.sendTwoWayCommand(method, rpcRequest, requestTimeout)`

But it only works from the dashboard after setting selected device in widget menu there. But there has to be some way to do the same in programatically?

Have tried the following without success:
```
ctx.defaultSubscription.targetDeviceId = "id";
ctx.defaultSubscription.targetDevice = {
  type: "device",
  deviceId: "id"
}
ctx.targetEntityId = {
  id: testDeviceId,
  entityType: "DEVICE"
};
ctx.targetEntityName = "name";

```

ChatGPT is unable to help as well 😂My last hope is the community!

**Comment by tobiasdecker** [comment]
Hello Sindre,
I'm currently trying to solve the same problem. Have you found a solution in the meantime? If so, it would be great if you or someone else could share it.
Thanks in advance!

**Comment by SindreLindahl** [comment]
Unfortunately not. Seems to be 2 layers the device id can be set. I can set it on a shallow layer like this: `ctx.defaultSubscription.targetDeviceId = "your-hard-coded-test-device-id";`. This will work in some cases, but not all.
